include/oneapi/dpl/internal/dynamic_selection_impl/sycl_backend.h

```diff
@@ -35,25 +37,89 @@ class sycl_backend
     using execution_resource_t = resource_type;
     using resource_container_t = std::vector<execution_resource_t>;
 
+    using report_clock_type = std::chrono::steady_clock;
+    using report_duration = std::chrono::duration<double, std::milli>;
+
+    static inline bool is_profiling_enabled = false;
   private:
-    class async_waiter
+    class async_waiter_base{
+        public:
+            virtual void wait() = 0;
+            virtual void report() = 0;
+            virtual bool is_complete() = 0;
+            virtual ~async_waiter_base() = default;
+    };
+
+    template<typename Selection>
+    class async_waiter : public async_waiter_base
     {
         sycl::event e_;
-
+        std::shared_ptr<Selection> s;
       public:
-        async_waiter(const sycl::event& e) : e_(e) {}
+        async_waiter(sycl::event e) : e_(e) {}
+        async_waiter(sycl::event e, std::shared_ptr<Selection> selection) : e_(e), s(selection) {}
+
+        async_waiter(async_waiter &w) : e_(w.e_), s(w.s) {}
         sycl::event
         unwrap()
         {
             return e_;
         }
+
         void
-        wait()
+        wait() override
         {
             e_.wait();
         }
+
+        void
+        report() override{
+            if constexpr (report_value_v<Selection, execution_info::task_time_t>){
+                cl_ulong time_start = e_.template get_profiling_info<sycl::info::event_profiling::command_start>();
+                cl_ulong time_end = e_.template get_profiling_info<sycl::info::event_profiling::command_end>();
+                if(s!=nullptr){
+                    s->report(execution_info::task_time, report_duration(time_end-time_start).count());
+                }
+            }
+
+
+        }
```

Comment:
```suggestion
        void
        report() override
        {
            if constexpr (report_value_v<Selection, execution_info::task_time_t>)
            {
                const auto time_start = e_.template get_profiling_info<sycl::info::event_profiling::command_start>();
                const auto time_end = e_.template get_profiling_info<sycl::info::event_profiling::command_end>();
                if (s != nullptr)
                {
                    // According to https://registry.khronos.org/SYCL/specs/sycl-2020/html/sycl-2020.html#_event_information_and_profiling_descriptors
                    // each profiling descriptor returns a 64-bit timestamp that represents the number of nanoseconds that have elapsed since some implementation-defined timebase.
                    const auto duration_in_ns = std::chrono::nanoseconds(time_end - time_start);

                    s->report(execution_info::task_time, std::chrono::duration_cast<report_duration>(duration_in_ns).count());
                }
            }
        }
```

Reply:
fixed